### ISSUE TITLE
NetStandardIDistributedCacheStringProvider breaks when using asp.net core 2.0

### ISSUE THREAD

**Issue by MortenChristiansen** [App-vNext/Polly.Caching.IDistributedCache#2]
When using this package with asp.net core 2.0 I get the following exception when the cache value is set:

```
Method not found: 'System.Threading.Tasks.Task Microsoft.Extensions.Caching.Distributed.DistributedCacheExtensions.SetStringAsync(Microsoft.Extensions.Caching.Distributed.IDistributedCache, System.String, System.String, Microsoft.Extensions.Caching.Distributed.DistributedCacheEntryOptions)'.
   at Polly.Caching.IDistributedCache.NetStandardIDistributedCacheStringProvider.PutAsync(String key, String value, Ttl ttl, CancellationToken cancellationToken, Boolean continueOnCapturedContext)
   at Polly.Caching.CacheEngine.<ImplementationAsync>d__1`1.MoveNext() in C:\projects\polly\src\Polly.Shared\Caching\CacheEngineAsync.cs:line 58
```

Asp.net core 2 references `Microsoft.Extensions.Caching.Abstractions` version 2.0.1, which has changed the signature in `DistributedCacheExtensions` from:

```
public static Task SetStringAsync(this IDistributedCache cache, string key, string value);
public static Task SetStringAsync(this IDistributedCache cache, string key, string value, DistributedCacheEntryOptions options);
```

to

`public static Task SetStringAsync(this IDistributedCache cache, string key, string value, DistributedCacheEntryOptions options = null);`

This means that that the code is built against a method that no longer exists (specifically `Task SetStringAsync(this IDistributedCache cache, string key, string value)`) causing it to fail at runtime. It should be a simple matter of bumping the dependency version and releasing a new version of the package.

**Comment by reisenberger** [comment]
Thank you @MortenChristiansen for identifying this so precisely.  Yes, we can fix by compiling against 2.0.

We will likely have to provide both a NetStandard 1.1 target and a NetStandard 2.0 target, to continue to support NetStandard 1.1-1.6  customers.

@joelhulen This sounds like the first case where we need a separate NetStandard 2.0 target because of 2.0 having a breaking (rather than just additive) change over 1.x.  

@MortenChristiansen There are a bunch of other packaging issues time-sensitive with the ASPNET Core 2.1 release schedule ( #3 ; https://github.com/App-vNext/Polly/issues/442 ), I'm just waiting on timing info for that before we know whether we'll fix this issue you've raised as part of that, or before.

---

Interestingly, the NetStandard doco states [there are no breaking changes between versions](https://docs.microsoft.com/en-us/dotnet/standard/net-standard#net-standard-versioning-rules); in this case that seems to be true at the lexical level but not the binary level.

**Comment by joelhulen** [comment]
So much for Standard being... um... standard... and backward-compatible. This is interesting, and thank you, @MortenChristiansen, for the detailed breakdown.

@reisenberger perhaps we should address the NetStandard 1.1 & 2.0 targeting with Polly v6.0? Then the changes to v6.0 packaging will be:

1. Having a single-signed package, doing away with the `Polly-Signed` lineage
2. Targeting NetStandard 1.1
3. Targeting NetStandard 2.0

What do you think?

**Comment by rynowak** [comment]
The caching issue is a breaking change in a major version update of our library. 

To clear something up... The `Microsoft.Extensions` packages **target** netstandard... but they are not part of the definition of netstandard. That's why they require an additional package dependency.  I realize that's not a super clear or helpful explanation - and it doesn't help resolve this situation.

We generally don't make breaking changes in the `Microsoft.Extensions` libraries (even in a major version) and I'm not sure how/why something like this was approved. 

You can see in the commit here that this was picked up by our tool that attempts to catch breaking changes https://github.com/aspnet/Caching/commit/d3bacc84e4e6954aaacfc43572d07aecb739fca7

**Edit:** I dug into this a bit more, and the reason why this was changed was that the core abstraction didn't accept a cancellation token. We had to break the interface, so it was seen as an unimportant matter to keep the extension methods exactly binary compatible since the interface changed.

------

So, if you want to live with the quirk that `netstandard1.1 -> Microsoft.Extensions.Caching.Abstractions 1.1.2` and `netstandard2.0 -> Microsoft.Extensions.Caching.Abstractions 2.0.X` that seems like a pragmatic choice. I can't think of any reason why anyone who is running on .NET461+ would not prefer the 2.0 version of these libraries.


---- 

As a personal aside, I'm a little vexxed by this guidance from https://docs.microsoft.com/en-us/dotnet/standard/net-standard

> In general, we recommend you to target the lowest version of .NET Standard possible. 

This is presented as a deep and meaningful decision, but really it seems pretty binary to me between whether you care about the NET45X family (netstandard1.2) or you don't (netstandard2.0). Unity just announced a release that supports netstandard2.0 so hopefully this will all start to get better quickly as the reasons to care about netstandard1.X disappear.

Sorry that this is causing you both pain, let me know if I can do anything else to help.

**Comment by MortenChristiansen** [comment]
Looking at the code again with fresh eyes it appears I made a small error in the description, though the result is the same. The problem is not that `DistributedCacheEntryOptions` has been made optional, but rather the optional `CancellationToken` parameter has been added. Just wanted to clarify that in case someone looked at the issue and got confused, though it is basically just what @rynowak said :)

**Comment by reisenberger** [comment]
@rynowak Thanks for the explanation: makes sense about `Microsoft.Extensions` packages not being part of the defined netstandard _surface_, good to keep in mind.

We'll go with a new Polly.Caching.IDistributedCache package with two targets: .NET Standard 1.1 and 2.0

@MortenChristiansen We're pushing through the repackaging/retargeting across the Polly repos, with [main Polly going first](https://github.com/App-vNext/Polly/pull/445), then [Polly.Extensions.Http](https://www.nuget.org/packages/Polly.Extensions.Http). Then we'll roll the same to the two existing caching packages, this Polly.Caching.IDistributedCache and Polly.Caching.MemoryCache. Stay tuned!

**Comment by reisenberger** [comment]
@MortenChristiansen, Polly and Polly.Extensions.Http repackaging released last week.  We are now looking at re-targeting on the two Polly.Caching packages (thanks @seanfarrow!), so more soon on this.

**EDIT:** #4 will update and fix this.

**Comment by reisenberger** [comment]
Closed by #6
